tests/xharness/TestTasks/MSBuildTask.cs

```diff
@@ -0,0 +1,116 @@
+			args.Add ($"/bl:{binlogPath}");
+			if (SpecifyPlatform)
+				args.Add ($"/p:Platform={projectPlatform}");
+			if (SpecifyConfiguration)
+				args.Add ($"/p:Configuration={projectConfiguration}");
+			args.Add (projectFile);
+			return args;
+		}
+
+		public MSBuildTask (string msbuildPath,
+							IProcessManager processManager,
+							IResourceManager resourceManager,
+							IEventLogger eventLogger,
+							IEnvManager envManager,
+							IErrorKnowledgeBase errorKnowledgeBase) : base (processManager, resourceManager, eventLogger, envManager)
+		{
+			this.msbuildPath = msbuildPath ?? throw new ArgumentNullException (nameof (msbuildPath));
+			this.errorKnowledgeBase = errorKnowledgeBase ?? throw new ArgumentNullException (nameof (errorKnowledgeBase));
+		}
+
+		public async Task<(TestExecutingResult ExecutionResult, string KnownFailure)> ExecuteAsync (string projectPlatform, 
+																									string projectConfiguration,
+																									string projectFile,
+																									IAcquiredResource resource,
+																									bool dryRun,
+																									ILog buildLog,
+																									ILog mainLog)
+		{
+			(TestExecutingResult ExecutionResult, string KnonwFailure) result = (TestExecutingResult.NotStarted, (string) null);
+			await RestoreNugetsAsync (buildLog, resource, useXIBuild: true);
+
+			using (var xbuild = new Process ()) {
+				xbuild.StartInfo.FileName = msbuildPath;
+				xbuild.StartInfo.Arguments = StringUtils.FormatArguments (GetToolArguments (projectPlatform, projectConfiguration, projectFile, buildLog));
+				EnviromentManager.SetEnvironmentVariables (xbuild);
+				xbuild.StartInfo.EnvironmentVariables ["MSBuildExtensionsPath"] = null;
+				EventLogger.LogEvent (buildLog, "Building {0} ({1})", TestName, Mode);
+				if (!dryRun) {
+					var timeout = TimeSpan.FromMinutes (60);
+					var processResult = await ProcessManager.RunAsync (xbuild, buildLog, timeout);
+					if (processResult.TimedOut) {
+						result.ExecutionResult = TestExecutingResult.TimedOut;
+						buildLog.WriteLine ("Build timed out after {0} seconds.", timeout.TotalSeconds);
+					} else if (processResult.Succeeded) {
+						result.ExecutionResult = TestExecutingResult.Succeeded;
+					} else {
+						result.ExecutionResult = TestExecutingResult.Failed;
+						if (errorKnowledgeBase.IsKnonwBuildIssue (buildLog, out result.KnonwFailure))
+							buildLog.WriteLine ($"Build has a known failure: '{result.KnonwFailure}'");
+					}
+				}
+				mainLog.WriteLine ("Built {0} ({1})", TestName, Mode);
+			}
+			return result;
+		}
+
+		async Task CleanProjectAsync (string project_file, string project_platform, string project_configuration, ILog log, ILog mainLog)
+		{
+			// Don't require the desktop resource here, this shouldn't be that resource sensitive
+			using (var xbuild = new Process ()) {
+				xbuild.StartInfo.FileName = Harness.XIBuildPath;
+				var args = new List<string> ();
```

Comment:
Could this part also use GetToolArguments somehow? They seem similar but maybe I overlooked something